service/dap/server.go

```diff
@@ -1541,10 +1528,103 @@ func (s *Server) onRestartRequest(request *dap.RestartRequest) {
 	s.sendNotYetImplementedErrorResponse(request.Request)
 }
 
-// onSetFunctionBreakpointsRequest sends a not-yet-implemented error response.
-// Capability 'supportsFunctionBreakpoints' is not set 'initialize' response.
 func (s *Server) onSetFunctionBreakpointsRequest(request *dap.SetFunctionBreakpointsRequest) {
-	s.sendNotYetImplementedErrorResponse(request.Request)
+	if s.noDebugProcess != nil {
+		s.sendErrorResponse(request.Request, UnableToSetBreakpoints, "Unable to set or clear breakpoints", "running in noDebug mode")
+		return
+	}
+	// TODO(polina): handle this while running by halting first.
+
+	// According to the spec we should "Replaces all existing function
+	// breakpoints with new function breakpoints." The simplest way is
+	// to clear all and then set all.
+	//
+	// TODO(polina): should we optimize this as follows?
+	// See https://github.com/golang/vscode-go/issues/163 for details.
+	// If a breakpoint:
+	// -- exists and not in request => ClearBreakpoint
+	// -- exists and in request => AmendBreakpoint
+	// -- doesn't exist and in request => SetBreakpoint
+
+	// Clear all existing function breakpoints in the file.
+	// Function breakpoints are set with the Name field set to be the
+	// function name. We cannot use FunctionName to determine this, because
+	// the debugger sets the function name for breakpoints.
+	if err := s.clearMatchingBreakpoints(func(bp *api.Breakpoint) bool { return bp.Name != "" }); err != nil {
+		s.sendErrorResponse(request.Request, UnableToSetBreakpoints, "Unable to set or clear breakpoints", err.Error())
+		return
+	}
+
+	// Set all requested breakpoints.
+	response := &dap.SetFunctionBreakpointsResponse{Response: *newResponse(request.Request)}
+	response.Body.Breakpoints = make([]dap.Breakpoint, len(request.Arguments.Breakpoints))
+	for i, want := range request.Arguments.Breakpoints {
+		spec, err := locspec.Parse(want.Name)
+		if err != nil {
+			response.Body.Breakpoints[i].Verified = false
+			response.Body.Breakpoints[i].Message = err.Error()
+			continue
+		}
+		if _, ok := spec.(*locspec.RegexLocationSpec); ok {
```

Comment:
Are we ok with users being able to say, for example, `+1` when setting a function breakpoint and that creates a breakpoint on the next line that the editor thinks is a function breakpoint?

Or to type in `main.go:15` as a function breakpoint.

Reply:
I updated this check to only allow NormalLocationSpecs, which should limit this to `func:line`, `file:line`. I documented that `file:line` is accepted in the tests.

While `func:line` is what this request is required to support, I'm not sure it is worth making sure that `file:line` is not also supported. On the other hand, address, offset, and regex locations do not make sense in the context of function breakpoints. We could update the locspec package to have a function that will only check for functions that match the spec and not files (this could also maybe be done in a separate change).

@polinasok Do you want have any thoughts on this?

Reply:
For the record: I don't have an opinion on this.

Reply:
My only issue is relative paths. We have been questioning their usage and resolution for everything else. As long as we clearly log/document what the situation is here, this should be fine.